### PR TITLE
refactor: add eventBus

### PR DESCRIPTION
**Pull request by xiyaowong**

### CODE REVIEW

File: src/cursor_manager.ts [hunk 1]
```diff
@@ -137,51 +139,32 @@ export class CursorManager implements Disposable, NeovimRedrawProcessable, Neovi
         }
     }
 
-    public handleRedrawBatch(batch: [string, ...unknown[]][]): void {
```
Comment (theol0403) [inline src/cursor_manager.ts:140]: I suppose these can be private now

**Review by theol0403 — APPROVED** [review]
Typescript wizardry!

**Comment by justinmk** [comment]
Nice. Can you explain the idea in the commit message, and possibly also in CONTRIBUTING.md or maybe an `architecture.md` doc. Or at least in `src/eventBus.ts`.

**Comment by theol0403** [comment]
@justinmk I believe there is little documentation due to the effort it takes for them to write english, but I think this PR is fairly straightforward - instead of having each manager implement an interface for handling events, which gets called by the main controller, now each manager subscribes to an event broadcast.

File: src/eventBus.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+export interface Event<T extends keyof EventsMapping = keyof EventsMapping> {
+    name: T;
+    data: EventsMapping[T];
+}
+
+export type EventBusData<T extends keyof EventsMapping> = EventsMapping[T];
+
+class EventBus implements Disposable {
+    private readonly emitter = new EventEmitter<Event>();
```
Comment (justinmk) [inline src/eventBus.ts:113]: Is the idea that all Nvim events are dispatched by this single `EventEmitter` ?

Comment (theol0403) [reply src/eventBus.ts:113]: Yeah I believe so. This will also let you write lua handlers for nvim events with the new api RFC

**Comment by justinmk** [comment]
> but I think this PR is fairly straightforward 

PRs in isolation are often easily digestible, but for maintenance / other contributors it's worth having a basic high-level description somewhere, which can be augmented over time.

In this case, just mentioning this one-sentence explanation somewhere is much better than nothing:

> instead of having each manager implement an interface for handling events, which gets called by the main controller, now each manager subscribes to an event broadcast.

That explanation could live in the module itself as well as an "architecture" section in CONTRIBUTING.md. Doesn't need to be fancy, the main thing is just to get something started so we have a place to add notes along the way.

**Comment by xiyaowong** [comment]
Merge?

File: src/eventBus.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+};
+
+export interface Event<T extends keyof EventsMapping = keyof EventsMapping> {
+    name: T;
+    data: EventsMapping[T];
+}
+
+export type EventBusData<T extends keyof EventsMapping> = EventsMapping[T];
+
+class EventBus implements Disposable {
```
Comment (justinmk) [inline src/eventBus.ts:112]: Here's what I was hoping for (and also a note in CONTRIBUTING.md):

```
/**
 * All Nvim events are dispatched by this single EventEmitter.
 * Components can subscribe to event broadcasts using `EventBus.on()`.
 */
```

**Pull request merged**